### PR TITLE
WIP Build dependencies from source

### PR DESCRIPTION
**Pull request by AxelNennker**
Signed-off-by: Axel Nennker <axel.nennker@telekom.de>

ZMQ's version 0.9.x now, like openssl, has the feature 'vendored' which compiles the source code for the given target.
Indy stayed at 0.8.3 because there are breaking changes in the API and send_str is now deprecated.
Please review networker.rs carefully.

So `cargo clean && PATH=$HOME/Android/standalone/bin:$PATH cargo build --target=aarch64-linux-android` now builds without complaints. Hurray.
Problem is that now the regular `cargo build` fails because some functions are defined twice.
Please see below.
The functions are defined in libsodium and in zeromq-src. Too bad.

@jean-airoldie did a great job creating this crate.
https://github.com/jean-airoldie/zeromq-src-rs
This crate's gitmodule points to https://github.com/zeromq/libzmq/blob/96b9eece158311095cb0e3db9e6d655773385b57/src/tweetnacl.c
The method crypto_stream_salsa20_xor is here:
https://github.com/zeromq/libzmq/blob/96b9eece158311095cb0e3db9e6d655773385b57/src/tweetnacl.c#L169

The other definition of the function crypto_stream_salsa20_xor is in libsodium-sys here which Indy inherits due to its dependency on sodiumoxide.
https://github.com/sodiumoxide/sodiumoxide/tree/master/libsodium-sys
The method crypto_stream_salsa20_xor is here:
https://github.com/jedisct1/libsodium/blob/927dfe8e2eaa86160d3ba12a7e3258fbc322909c/src/libsodium/include/sodium/crypto_stream_salsa20.h#L41

I need to find a way to undefine ZMQ_USE_TWEETNACL in zeromq-src, I guess.
Its configure has a `--with-libsodium` option but then I need zeromq-src to find libsodium.a where sodiumoxide created it. Tricky.

Kind of strange that this works for Android but not for Ubuntu. Stoked.

```  = note: /home/ignisvulpis/development/hyperledger/indy-sdk/libindy/target/debug/deps/liblibsodium_sys-154ac4e7141d08f8.rlib(libsodium_la-stream_salsa20.o): In function `crypto_stream_salsa20':
          stream_salsa20.c:(.text+0x30): multiple definition of `crypto_stream_salsa20'
          /home/ignisvulpis/development/hyperledger/indy-sdk/libindy/target/debug/deps/libzmq_sys-01baefcf1c668911.rlib(tweetnacl.c.o):/home/ignisvulpis/.cargo/registry/src/github.com-1ecc6299db9ec823/zeromq-src-0.1.6+4.3.1/vendor/src/tweetnacl.c:197: first defined here
          /home/ignisvulpis/development/hyperledger/indy-sdk/libindy/target/debug/deps/liblibsodium_sys-154ac4e7141d08f8.rlib(libsodium_la-stream_salsa20.o): In function `crypto_stream_salsa20_xor':
          stream_salsa20.c:(.text+0x50): multiple definition of `crypto_stream_salsa20_xor'
          /home/ignisvulpis/development/hyperledger/indy-sdk/libindy/target/debug/deps/libzmq_sys-01baefcf1c668911.rlib(tweetnacl.c.o):tweetnacl.c:(.text.crypto_stream_salsa20_xor+0x0): first defined here
          /home/ignisvulpis/development/hyperledger/indy-sdk/libindy/target/debug/deps/liblibsodium_sys-154ac4e7141d08f8.rlib(libsodium_la-verify.o): In function `crypto_verify_16':
          verify.c:(.text+0x30): multiple definition of `crypto_verify_16'
          /home/ignisvulpis/development/hyperledger/indy-sdk/libindy/target/debug/deps/libzmq_sys-01baefcf1c668911.rlib(tweetnacl.c.o):tweetnacl.c:(.text.crypto_verify_16+0x0): first defined here
          /home/ignisvulpis/development/hyperledger/indy-sdk/libindy/target/debug/deps/liblibsodium_sys-154ac4e7141d08f8.rlib(libsodium_la-verify.o): In function `crypto_verify_32':
          verify.c:(.text+0xb0): multiple definition of `crypto_verify_32'
          /home/ignisvulpis/development/hyperledger/indy-sdk/libindy/target/debug/deps/libzmq_sys-01baefcf1c668911.rlib(tweetnacl.c.o):tweetnacl.c:(.text.crypto_verify_32+0x0): first defined here
          /home/ignisvulpis/development/hyperledger/indy-sdk/libindy/target/debug/deps/liblibsodium_sys-154ac4e7141d08f8.rlib(libsodium_la-randombytes.o): In function `randombytes_close':
          randombytes.c:(.text+0x210): multiple definition of `randombytes_close'
          /home/ignisvulpis/development/hyperledger/indy-sdk/libindy/target/debug/deps/libzmq_sys-01baefcf1c668911.rlib(tweetnacl.c.o):tweetnacl.c:(.text.randombytes_close+0x0): first defined here
          /home/ignisvulpis/development/hyperledger/indy-sdk/libindy/target/debug/deps/liblibsodium_sys-154ac4e7141d08f8.rlib(libsodium_la-randombytes.o): In function `randombytes':
          randombytes.c:(.text+0x240): multiple definition of `randombytes'
          /home/ignisvulpis/development/hyperledger/indy-sdk/libindy/target/debug/deps/libzmq_sys-01baefcf1c668911.rlib(tweetnacl.c.o):tweetnacl.c:(.text.randombytes+0x0): first defined here
          /home/ignisvulpis/development/hyperledger/indy-sdk/libindy/target/debug/deps/liblibsodium_sys-154ac4e7141d08f8.rlib(libsodium_la-core.o): In function `sodium_init':
          core.c:(.text+0x50): multiple definition of `sodium_init'
          /home/ignisvulpis/development/hyperledger/indy-sdk/libindy/target/debug/deps/libzmq_sys-01baefcf1c668911.rlib(tweetnacl.c.o):tweetnacl.c:(.text.sodium_init+0x0): first defined here
          collect2: error: ld returned 1 exit status
```

### CODE REVIEW

**Comment by AxelNennker** [comment]
opened https://github.com/jean-airoldie/zeromq-src-rs/issues/8

**Comment by AxelNennker** [comment]
https://github.com/jean-airoldie fixed this in zmq-src-rs already
https://github.com/jean-airoldie/zeromq-src-rs/issues/8

He also created a PR to sodiumoxide that makes the includes directory and library directory, that has libsodium.a in it, available to other crates. If this is merged this PR should work as well for all platforms.
https://github.com/sodiumoxide/sodiumoxide/pull/344

Could somebody please try out iOS and windows? Android works now, although I do not know why, so iOS might work as well or provide more information.

File: libindy/build.rs [hunk 1]
```diff
@@ -32,34 +32,10 @@ fn main() {
...
-        let zmq = match env::var("LIBZMQ_LIB_DIR") {
-            Ok(val) => val,
-            Err(..) => match env::var("LIBZMQ_PREFIX") {
-                Ok(dir) => Path::new(&dir[..]).join("lib").to_string_lossy().into_owned(),
-                Err(..) => panic!("Missing required environment variables LIBZMQ_PREFIX or LIBZMQ_LIB_DIR")
-            }
-        };
-
-        println!("cargo:rustc-link-search=native={}", openssl);
+        // statically link libraries
```
Comment (mikelodder7) [inline libindy/build.rs]: Do you want to remove the directory hints? Not sure about this.

File: libindy/src/services/pool/networker.rs [hunk 2]
```diff
@@ -343,13 +343,10 @@ impl PoolConnection {
...
         let s = ctx.socket(zmq::SocketType::DEALER)?;
-        s.set_identity(key_pair.public_key.as_bytes())?;
+        s.set_identity(&key_pair.public_key)?;
         s.set_curve_secretkey(&key_pair.secret_key)?;
         s.set_curve_publickey(&key_pair.public_key)?;
-        s.set_curve_serverkey(
-            zmq::z85_encode(self.public_key.as_slice())
-                .to_indy(IndyErrorKind::InvalidStructure, "Can't encode server key as z85")? // FIXME: review kind
-                .as_str())?;
+        s.set_curve_serverkey(self.public_key.as_slice())?;
```
Comment (mikelodder7) [inline libindy/src/services/pool/networker.rs]: The z85 just shrinks the number of bytes. So what is the reason this was removed?

File: libindy/build.rs (hunk 1 above)
Comment (AxelNennker) [reply libindy/build.rs]: On Android all dependencies are now build from source. The -src crates set these directory hints.
https://github.com/alexcrichton/openssl-src-rs
https://github.com/jean-airoldie/zeromq-src-rs
WIP https://github.com/sodiumoxide/sodiumoxide/pull/344

File: libindy/Cargo.toml [hunk 3]
```diff
@@ -1,14 +1,7 @@
 [package]
 name = "libindy"
-version = "1.11.0"
-authors = [
-  "Sergej Pupykin <sergej.pupykin@dsr-company.com>",
-  "Vyacheslav Gudkov <vyacheslav.gudkov@dsr-company.com>",
-  "Artem Ivanov <artem.ivanov@dsr-company.com>",
-  "Evgeniy Razinkov <evgeniy.razinkov@dsr-company.com.ru>",
-  "Kirill Neznamov <kirill.neznamov@dsr-company.com>",
-  "Sergey Minaev <sergey.minaev@dsr-company.com>"
```
Comment (kpp) [inline libindy/Cargo.toml]: Woah...

**Comment by AxelNennker** [comment]
Build failed with 'java.io.IOException: No space left on device'.
Please restart after this issue is fixed

**Pull request closed without merging**